hi-ml-histopathology/src/histopathology/datamodules/base_module.py

```diff
@@ -150,40 +201,139 @@ def _load_dataset(self, tiles_dataset: TilesDataset, stage: str, shuffle: bool)
+    def _get_dataloader(
+        self, dataset: TilesDataset, stage: str, shuffle: bool, **dataloader_kwargs: Any
+    ) -> DataLoader:
+        transformed_bag_dataset = self._load_dataset(dataset, stage=stage, shuffle=shuffle)
         bag_dataset: BagDataset = transformed_bag_dataset.data  # type: ignore
         generator = bag_dataset.bag_sampler.generator
-        return DataLoader(transformed_bag_dataset, batch_size=self.batch_size,
-                          collate_fn=multibag_collate, shuffle=shuffle, generator=generator,
-                          **dataloader_kwargs)
+        return DataLoader(
+            transformed_bag_dataset,
+            batch_size=self.batch_size,
+            collate_fn=multibag_collate,
+            shuffle=shuffle,
+            generator=generator,
+            pin_memory=False,  # disable pinning as loaded data may already be on GPU
+            **dataloader_kwargs,
+        )
 
-    def train_dataloader(self) -> DataLoader:
-        return self._get_dataloader(self.train_dataset, 'train', shuffle=True, **self.dataloader_kwargs)
 
-    def val_dataloader(self) -> DataLoader:
-        return self._get_dataloader(self.val_dataset, 'val', shuffle=True, **self.dataloader_kwargs)
+class SlidesDataModule(HistoDataModule):
+    """
+    Base class to load the slides of a dataset as train, val, test sets
+    Args:
+        level: the level number, or list of level numbers (default=0)
+        tile_count: number of tiles to extract, if None extracts all non-background tiles
+            Defaults to ``None``.
+        tile_size: size of the square tile
+            Defaults to ``224``.
+        step: step size to create overlapping tiles
+            Defaults to ``None`` (same as tile_size)
+        random_offset: Randomize position of the grid, instead of starting from the top-left corner
+            Defaults to ``False``.
+        pad_full: pad image to the size evenly divisible by tile_size
+            Defaults to ``False``.
+        background_val: the background constant (e.g. 255 for white background)
+            Defaults to ``255``.
+        filter_mode: mode must be in ["min", "max", "random"]. If total number of tiles is more than tile_size,
+            then sort by intensity sum, and take the smallest (for min), largest (for max) or random (for random) subset
+            Defaults to ``min`` (which assumes background is high value)
+    """
 
-    def test_dataloader(self) -> DataLoader:
-        return self._get_dataloader(self.test_dataset, 'test', shuffle=True, **self.dataloader_kwargs)
+    def __init__(
+        self,
+        level: Optional[int] = 0,
+        tile_count: Optional[int] = None,
+        tile_size: Optional[int] = 224,
+        step: Optional[int] = None,
+        random_offset: Optional[bool] = False,
+        pad_full: Optional[bool] = False,
+        background_val: Optional[int] = 255,
+        filter_mode: Optional[str] = "min",
```

Comment:
I would default this to `random`, this way we will be sampling a different set of tiles at each step, otherwise we would need to sample many more tiles per slide to get a representative sample